active/0020-node-evacuation.md

```diff
@@ -0,0 +1,356 @@
+Eviction Agent exposes two endpoints: `/node_eviction/status` and `/node_eviction/heath_check`.
+
+#### Algorithm
+
+##### `disabled` state
+
+The disabled state is the default. In the `disabled` state Eviction Agent does not influence the node anyhow.
+
+HTTP responses are the following:
+
+Request:
+```
+GET /api/v4/node_eviction/status
+```
+
+Response:
+```
+HTTP/1.1 200 OK
+
+{"status": "disabled"}
+```
+
+Request:
+```
+GET /api/v4/node_eviction/heath_check
+```
+
+Response:
+```
+HTTP/1.1 200 OK
+```
+
+The internal API behaviour is the following:
+
+```erlang
+{disabled,
+   #{connected_channels := ConnectedCount,
+     total_channels := TotalCount,
+     sessions := SessonCount }} = emqx_eviction_agent:status(Node).
+```
+
+From the `disabled` state Eviction Agent can be enabled either temporarily or permanently.
+
+##### Enable temporarily
+
+Enabling Eviction Agent temporarily requires two arguments:
+* Server reference for [Server redirect](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901255)
+* Pid of the coordinator process (the caller).
+
+When being enabled temporarily, Eviction Agent:
+* installs prohibiting hooks: `client.connect` and `client.connack`;
+* monitors the Pid of the coordinator.
+
+The internal API is the following:
+
+```erlang
+{ok,
+    #{connected_channels := ConnectedCount,
+      total_channels := TotalCount,
+      sessions := SessonCount }} = emqx_eviction_agent:enable_temporarily(Node, ["host1:port1", "host2:port2"]).
+```
+
+##### Enable permanently
+
+Enabling Eviction Agent permanently requires a single argument:
+* Server reference for [Server redirect](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901255)
+
+When being enabled permanently, Eviction Agent:
+* installs prohibiting hooks: `client.connect` and `client.connack`;
+* creates a local file containing metadata (Server reference) as an indicator of the enabled state.
+
+The internal API is the following:
+
+```erlang
+{ok,
+    #{connected_channels := ConnectedCount,
+      total_channels := TotalCount,
+      sessions := SessonCount }} = emqx_eviction_agent:enable_permanently(Node, ["host1:port1", "host2:port2"]).
+```
+
+##### Enabled (`enabled_permanently`/`enabled_temporarily`)
+
+When enabled, Eviction Agent has installed hooks. Their logic is:
+* `client.connect` hook returns `CONNACK_SERVER` (for MQTT 3) or `RC_USE_ANOTHER_SERVER` (for MQTT 5) error.
+* `client.connack` sets `'Server-Reference'` property to the specified servers `host1:port1 host2:port2 ...` for MQTT 5
+clients acked with `RC_USE_ANOTHER_SERVER`. `'Server-Reference'` property is not set if servers are not specified or
+for MQTT 3 clients.
+
+HTTP responses are the following:
+
+Request:
+```
+GET /api/v4/node_eviction/status
+```
+
+Response:
+```
+HTTP/1.1 200 OK
+
+{
+    "status": "enabled_permanently", # enabled_temporarily
+    "stats": {
+        "connected": 25,
+        "total": 100,
+        "sessions": 50
+    }
+}
```

Comment:
maybe something like this:
```suggestion
{
    "status": "enabled",
    "goal" 0,
    "stats": {
        "initial_connected": 25,
        "initial_sessions": 50,
        "current_connected": 25,
        "current_sessions": 50
    }
}
```
`goal` being `0` means evacuate all connections from this node.
`goal` can be extended in the future, e.g. `cluster_balance` to indicate a dynamic goal, which should be re-calculated from time to time.